lib/collection/src/collection/clean.rs

```diff
@@ -0,0 +1,318 @@
+            let filter = shard_holder
+                .hash_ring_filter(shard_id)
+                .expect("hash ring filter");
+            let filter = Filter::new_must_not(Condition::CustomIdChecker(Arc::new(filter)));
+            let mut ids = match shard
+                .scroll_by(
+                    offset,
+                    CLEAN_BATCH_SIZE + 1,
+                    &false.into(),
+                    &false.into(),
+                    Some(&filter),
+                    None,
+                    true,
+                    None,
+                    None,
+                )
+                .await
+            {
+                Ok(batch) => batch.into_iter().map(|entry| entry.id).collect::<Vec<_>>(),
+                Err(err) => {
+                    break ShardCleanStatus::Failed(format!(
+                        "Failed to read points to delete from shard: {err}"
+                    ));
+                }
+            };
+
+            // Check if cancelled
+            if cancel.is_cancelled() {
+                break ShardCleanStatus::Cancelled;
+            }
+
+            // Update offset for next batch
+            offset = (ids.len() > CLEAN_BATCH_SIZE).then(|| ids.pop().unwrap());
+            let last_batch = offset.is_none();
+
+            // Delete points from local shard
+            let delete_operation =
+                OperationWithClockTag::from(CollectionUpdateOperations::PointOperation(
+                    crate::operations::point_ops::PointOperations::DeletePoints { ids },
+                ));
+            if let Err(err) = shard.update_local(delete_operation, true).await {
+                break ShardCleanStatus::Failed(format!(
+                    "Failed to delete points from shard: {err}"
+                ));
+            }
+
+            // Finish if this was the last batch
+            if last_batch {
+                break ShardCleanStatus::Done;
+            }
+        };
+
+        let _ = sender.send(status);
+    }
+}
+
+impl Collection {
+    pub async fn cleanup_local_shard(
+        &self,
+        shard_id: ShardId,
+        wait: bool,
+        timeout: Option<Duration>,
+    ) -> CollectionResult<UpdateResult> {
+        let status = self
+            .shard_clean_tasks
+            .clean_and_await(&self.shards_holder, shard_id, wait, timeout)
+            .await?;
+
+        Ok(UpdateResult {
+            operation_id: None,
+            status,
+            clock_tag: None,
+        })
+    }
+
+    pub(super) fn invalidate_clean_local_shard(&self, shard_id: ShardId) {
+        self.shard_clean_tasks.invalidate(shard_id);
+    }
+
+    pub fn clean_local_shards_statuses(&self) -> HashMap<ShardId, ShardCleanStatusTelemetry> {
+        self.shard_clean_tasks
+            .statuses()
+            .into_iter()
+            .map(|(shard_id, status)| (shard_id, status.into()))
+            .collect()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(super) enum ShardCleanStatus {
+    Started,
```

Comment:
maybe we could track the starting time in this variant?

Reply:
Can do!

I actually plan to add a `Progress(something)` state which keeps being bumped until we eventually finish. I think I like that better as it'd show it's actually doing something.

If we implement the above we'd loose the Started variant almost immediately, also loosing the start time. So based on that I don't think it's worth it to include a starting time.

Reply:
I've just implemented the progress signal:

https://github.com/user-attachments/assets/49994794-4ca3-40fe-b2b0-0406d1e9ff4a